### PR TITLE
Allow conversion of ANSI/Terminal colors to IRC colors

### PR DESCRIPTION
**Pull request by ammaraskar**

### CODE REVIEW

**Comment by solarkennedy** [comment]
Test are failing, looks like there is also some bitrot on the gems and junk.

I'll help you fix those up as well if you want.

Ruby code looks good to me?

**Comment by ammaraskar** [comment]
Hey @bobtfish or @keymone, can I get a review on this please. kwa wants someone who actually knows ruby :)

**Comment by somic** [comment]
If tests pass and cover all code paths, I think it's ok to ship.

**Comment by solarkennedy** [comment]
Bonus points for extracting this code and publishing your first RubyGem!

File: files/nodebot.rb [hunk 1]
```diff
@@ -57,5 +60,101 @@ def handle
...
+ANSI_TO_IRC_BACK = {}
+# Formatting characters
+ANSI_TO_IRC_FMT = {
+  0  => 0xf,  # reset
+  1  => 0x2,  # bold
+  4  => 0x1f, # underline
+  7  => 0x16, # reverse mode
+}
+# Generate background colors
+ANSI_TO_IRC_COLORS.each { |k, v|
```
Comment (keymone) [inline files/nodebot.rb]: multiline blocks should be wrapped in `do` / `end` and this whole thing belongs closer to where `ANSI_TO_IRC_BACK` is defined

File: files/nodebot.rb [hunk 2]
```diff
@@ -57,5 +60,101 @@ def handle
...
+  # No ANSI sequences :(
+  if !token
+     return message
+  end
+
+  while token
+    # Remove the ANSI escape sequence
+    # and append whatever is left to
+    # the output
+    output += token[0...-1]
```
Comment (keymone) [inline files/nodebot.rb]: here and in other places `<<` is preferred way to append something to a string

File: files/nodebot.rb [hunk 3]
```diff
@@ -57,5 +60,101 @@ def handle
...
+
+    next_byte = scanner.get_byte()
+    if next_byte >= "0x40" and next_byte <= "0x7E"
```
Comment (keymone) [inline files/nodebot.rb]: `&&` is preferred due to `and`'s precedence rules

File: files/nodebot.rb [hunk 4]
```diff
@@ -57,5 +60,101 @@ def handle
...
+      sequence = scanner.scan_until(/[\x40-\x7E]/)
+      # Discard the final byte
+      sequence = sequence[0...-1]
+      # Split by semicolon
+      sequence = sequence.split(";")
+
+      foreground = nil
+      background = nil
+      formatting = nil
+      sequence.each { |ansi_code|
```
Comment (keymone) [inline files/nodebot.rb]: `do` / `end`

File: files/nodebot.rb [hunk 5]
```diff
@@ -57,5 +60,101 @@ def handle
...
+          ansi_code = Integer(ansi_code) rescue nil
+          if ansi_code
```
Comment (keymone) [inline files/nodebot.rb]: idiomatic thing here is to do `next unless ansi_code`

File: files/nodebot.rb [hunk 6]
```diff
@@ -57,5 +60,101 @@ def handle
...
+            foreground ||= ANSI_TO_IRC_COLORS[ansi_code]
+            background ||= ANSI_TO_IRC_BACK[ansi_code]
+            formatting ||= ANSI_TO_IRC_FMT[ansi_code]
+          end
+      }
+
+      if !foreground.nil?
```
Comment (keymone) [inline files/nodebot.rb]: here and in other places: `if !foo.nil?` is identical to just `if foo` unless you want this condition to eval to `true` for `false` value of `foo`, which you don't in this case

File: files/nodebot.rb [hunk 7]
```diff
@@ -57,5 +60,101 @@ def handle
...
+      end
+    else
+      # not a color, not much we can do about this
+      scanner.scan_until(/[\x40-\x7E]/)
+    end
+
+    token = scanner.scan_until(ANSI_START)
+  end
+  # Consume anything left
+  output += scanner.rest()
```
Comment (keymone) [inline files/nodebot.rb]: `output << scanner.rest()` and drop the return as `<<` returns the modified string

**Comment by keymone** [comment]
late to the party. just ruby's coding guidelines comments from my side, nothing wrong with code, nice work!

**Pull request merged**